### PR TITLE
chore: check genesis block during startup

### PR DESCRIPTION
**Pull request by Cifko**
Description
---
Check the genesis block from the binary is the same as the one in the DB.

### CODE REVIEW

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 1]
```diff
@@ -230,6 +230,14 @@ where B: BlockchainBackend
             txn.set_horizon_data(kernel_sum, utxo_sum);
             blockchain_db.write(txn)?;
             blockchain_db.store_pruning_horizon(config.pruning_horizon)?;
+        } else if blockchain_db.get_chain_metadata()?.pruning_horizon() == 0 {
+            // Check the genesis block in the DB.
+            if !blockchain_db.block_exists(genesis_block.accumulated_data().hash.clone())? {
+                error!(target: LOG_TARGET, "Stored genesis block mismatch with the binary");
```
Comment (delta1) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: Please add the expected hash of the genesis block, and the hash of the block in the db to this log

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 2]
```diff
@@ -230,6 +230,14 @@ where B: BlockchainBackend
...
+                return Err(ChainStorageError::CorruptedDatabase(
+                    "Genesis block mismatch.".to_string(),
```
Comment (delta1) [inline base_layer/core/src/chain_storage/blockchain_database.rs]:
```suggestion
                    "Genesis block in database does not match the supplied genesis block in the code! Please delete and resync your blockchain database.".into(),
```

**Review by delta1 — COMMENTED** [review]
Nice looking good, 2 small comments

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 2 above)
Comment (Cifko) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: Thanks. I'm really not good with wording errors :-)

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 3]
```diff
@@ -230,6 +230,22 @@ where B: BlockchainBackend
             txn.set_horizon_data(kernel_sum, utxo_sum);
             blockchain_db.write(txn)?;
             blockchain_db.store_pruning_horizon(config.pruning_horizon)?;
+        } else if blockchain_db.get_chain_metadata()?.pruning_horizon() == 0 {
+            // Check the genesis block in the DB.
+            if !blockchain_db.block_exists(genesis_block.accumulated_data().hash.clone())? {
+                error!(
+                    target: LOG_TARGET,
+                    "Genesis block in database does not match the supllied genesis block in the code! Hash in the \
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs]:
```suggestion
                    "Genesis block in database does not match the supplied genesis block in the code! Hash in the \
```

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 4]
```diff
@@ -230,6 +230,22 @@ where B: BlockchainBackend
             txn.set_horizon_data(kernel_sum, utxo_sum);
             blockchain_db.write(txn)?;
             blockchain_db.store_pruning_horizon(config.pruning_horizon)?;
+        } else if blockchain_db.get_chain_metadata()?.pruning_horizon() == 0 {
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: The hash should be the same for pruned or archival nodes, so can't we exclude this check?

Comment (Cifko) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: Removed.

**Review by delta1 — APPROVED** [review]

**Pull request merged**